module/state_synchronization/indexer/range.go

```diff
@@ -0,0 +1,90 @@
+package indexer
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+var (
+	ErrIndexValue    = fmt.Errorf("invalid index value")
+	ErrIndexBoundary = fmt.Errorf("the value is not within the indexed interval")
+)
+
+// SequentialIndexRange contains an interval bounded by two whole numbers, and it can only be increased
+// by incrementing the last value by exactly one, or if the same value is provide we ignore it to make the
+// mutation idempotent. This can be useful when we want to index height range.
+// SequentialIndexRange makes all the mutations atomic, so it can be considered concurrent-safe.
+type SequentialIndexRange struct {
+	first uint64 // we never allow changing first so no need to be atomic
```

Comment:
Maybe we won';t neeed to keep track of `first` at all? if the goal here is to record `Height` that we need to sync sequentially, we can just increment and keep track of that one height. I see thay `first` is just so we can use `Contained` below, but `pebble.Get` already performs the same check and provides the same error, we won't need to perform the same check twice.

Reply:
I think others can pitch in, but I really don't like relying on the storage layer to be the only one implementing this domain logic of the indexer. I see storage having an extra check (my opinion is it wouldn't even have to have it), but I personally don't think it's a good idea to have domain logic live inside storage layer, first of all, that then gets duplicated between different storage implementations (what if we have different db in future exactly like we now are introducing pebble), it complicates testing (mocks needs to have this checks somehow), it hides this explicit logic operation behind an API that doesn't explain that (Store/Get - this methods don't imply this restrictions at all), I could go on with design reasons for my decisions, but since it's 1 vs 1 in this case other should pitch in to see where the scale tips. cc @peterargue

Reply:
https://github.com/onflow/flow-go/pull/4653#discussion_r1335681180